pkg/volumes/alicloud/volumes.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/golang/glog"
+	"kope.io/etcd-manager/pkg/volumes"
+)
+
+const (
+	// We could query at most 50 disks at a time on Alicloud
+	PageSizeLarge = 50
+)
+
+type AlicloudVolumes struct {
+	mutex sync.Mutex
+
+	matchTagKeys []string
+	matchTags    map[string]string
+	nameTag      string
+	clusterName  string
+
+	deviceMap  map[string]string
+	ecs        *ecs.Client
+	instanceID string
+	localIP    string
+	metadata   *ECSMetadata
+	az         string
+	region     string
+}
+
+var _ volumes.Volumes = &AlicloudVolumes{}
+
+// NewAlicloudVolumes returns a new aws volume provider
+func NewAlicloudVolumes(clusterName string, volumeTags []string, nameTag string) (*AlicloudVolumes, error) {
+	a := &AlicloudVolumes{
+		clusterName: clusterName,
+		deviceMap:   make(map[string]string),
+		matchTags:   make(map[string]string),
+		nameTag:     nameTag,
+	}
+
+	for _, volumeTag := range volumeTags {
+		tokens := strings.SplitN(volumeTag, "=", 2)
+		if len(tokens) == 1 {
+			a.matchTagKeys = append(a.matchTagKeys, tokens[0])
+		} else {
+			a.matchTags[tokens[0]] = tokens[1]
+		}
+	}
+
+	a.metadata = NewECSMetadata()
+
+	var err error
+
+	a.region, err = a.metadata.GetMetadata("region-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for region): %v", err)
+	}
+
+	a.az, err = a.metadata.GetMetadata("zone-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for az): %v", err)
+	}
+
+	a.instanceID, err = a.metadata.GetMetadata("instance-id")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for instance-id): %v", err)
+	}
+
+	a.localIP, err = a.metadata.GetMetadata("private-ipv4")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for private-ipv4): %v", err)
+	}
+
+	ramRole, err := a.metadata.GetMetadata("ram/security-credentials/")
+	if err != nil {
+		return nil, fmt.Errorf("error querying ecs metadata (for RAM role): %v", err)
+	}
+
+	config := alicloud.NewConfig()
+
+	credential := credentials.NewEcsRamRoleCredential(ramRole)
+	a.ecs, err = ecs.NewClientWithOptions(a.region, config, credential)
+
+	return a, nil
+}
+
+// AttachVolume attaches the specified volume to this instance, returning the mountpoint & nil if successful
+func (a *AlicloudVolumes) AttachVolume(volume *volumes.Volume) error {
+	volumeID := volume.ProviderID
+
+	if volume.LocalDevice == "" {
+		attachDiskReq := ecs.CreateAttachDiskRequest()
+		attachDiskReq.RegionId = a.region
+		attachDiskReq.InstanceId = a.instanceID
+		attachDiskReq.DiskId = volumeID
+
+		attachResponse, err := a.ecs.AttachDisk(attachDiskReq)
+		if err != nil {
+			return fmt.Errorf("Unable to attach volume %q: %v", volumeID, err)
+		}
+
+		glog.V(2).Infof("AttachVolume request returned %v", attachResponse)
+
+		// Attaching volume takes time.
+		time.Sleep(5)
+	}
+
+	for {
```

Comment:
endless loop is not imo good idea here. We should have timeout for instance 5 minutes

Reply:
Added a 5 min timeout to the for loop. and rebased.